src/platform_impl/linux/wayland/event_loop.rs

```diff
@@ -172,6 +265,39 @@ impl<T: 'static> EventLoop<T> {
                                 .unwrap(),
                         )
                     }
+                    if interface == "zwp_pointer_constraints_v1" {
+                        let pointer_constraints_proxy = registry
+                            .bind(version, id, move |pointer_constraints| {
+                                pointer_constraints.implement_closure(|_, _| (), ())
+                            })
+                            .unwrap();
+
+                        *seat_manager.pointer_constraints_proxy.borrow_mut() =
+                            Some(pointer_constraints_proxy);
+                    }
+                    if interface == "wl_shm" {
+                        let shm: WlShm = registry
+                            .bind(version, id, move |shm| shm.implement_closure(|_, _| (), ()))
+                            .unwrap();
+
+                        (*shm_cell.borrow_mut()) = Some(shm);
+                    }
+                    if interface == "wl_compositor" {
+                        let compositor: WlCompositor = registry
+                            .bind(version, id, move |compositor| {
+                                compositor.implement_closure(|_, _| (), ())
+                            })
+                            .unwrap();
+                        (*compositor_cell.borrow_mut()) = Some(compositor);
+                    }
+
+                    if compositor_cell.borrow().is_some() && shm_cell.borrow().is_some() {
+                        let compositor = compositor_cell.borrow_mut().take().unwrap();
+                        let shm = shm_cell.borrow_mut().take().unwrap();
+                        let auto_themer = AutoThemer::init(None, compositor, &shm);
+                        cursor_manager2.borrow_mut().set_auto_themer(auto_themer);
+                    }
+
```

Comment:
The `Environment` callback is mainly targeted for globals that can appear after the initial setup (like seats or outputs). For functionality globals that are either always present or absent depending on whether the server supports it, it'll probably be easier to try & get them though `[env.manager.instanciate_exact(..)`](https://docs.rs/wayland-client/0.23.5/wayland_client/struct.GlobalManager.html#method.instantiate_exact) once the `Environment` has been created.

Reply:
I would like to use the environment to obtain the compositor and shm, but I need the AutoThemer when the pointers are registered. This happens during initialization of the Environment.

Reply:
Ah right. Well, this is the kind of dependencies between globals that could probably be better handled by SCTK directly I guess.